### PR TITLE
MAINT: Use ruff-format instead of running black in pre-commit

### PR DESCRIPTION
**Pull request by MridulS**
This resolves https://github.com/astropy/astropy/issues/15757 as https://github.com/astropy/astropy-APEs/pull/90 has been accepted.

https://github.com/astropy/astropy/commit/c69079e866c0cf394759bd7d1ffcda1a0e48cb0c has the changes caused by this.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
I was curious to see how much this saved, so I looked at the recent history on pre-commit.ci
The run for this PR isn't hard to spot !
Screenshot 2024-02-13 at 12 48 07

**Review by neutrinoceros — COMMENTED** [review]
I think it is customary to log large formatting commits in `.git-blame-ignore-revs` so they don't show up in git blame

**Comment by pllim** [comment]
We do have that file from when we initially implemented https://github.com/astropy/astropy-APEs/blob/main/APE20.rst but since then, we have not been adding follow-up ruff/black updates to it. Regardless, I don't think we can do this until we get the final merge commit after merging this PR? Anyways, if you want something added to https://github.com/astropy/astropy/blob/main/.git-blame-ignore-revs , either we do follow-up PR or I can do it directly on `main` as needed. Thanks!

**Review by eerovaher — CHANGES_REQUESTED** [review]
1. #15998 should be merged before we start working on this.
2. We have configured Black to ignore `docs/` and `examples/`. I think it would be a good idea to start formatting Python code in those directories too, but not in this pull request.
3. If we are not using Black anymore then all the Black configuration in `pyproject.toml` should be converted to equivalent Ruff formatter configuration.
4. Our developer documentation needs to be updated.

**Review by WilliamJamieson — APPROVED** [review]
The modeling changes look like an actual improvement.

**Review by larrybradley — APPROVED** [review]
`stats` looks fine.

**Review by taldcroft — APPROVED** [review]
Fine for time.

**Review by mcara — APPROVED** [review]
OK for WCS

**Review by nstarman — APPROVED** [review]
cosmology & units both LGTM!
Also +1 to @eerovaher's list.

**Comment by MridulS** [comment]
> 2. We have configured Black to ignore `docs/` and `examples/`. I think it would be a good idea to start formatting Python code in those directories too, but not in this pull request.

I will open a new PR for that. Copied over the black exclude list to ruff format too.
 
> 3. If we are not using Black anymore then all the Black configuration in `pyproject.toml` should be converted to equivalent Ruff formatter configuration.

Done.
 
> 4. Our developer documentation needs to be updated.

I have updated it in the places I saw black pop up. I have still put in links to the OG black style guide as ruff format wants to use that as the base and not deviate too much from it and the black docs explain it much better than ruff's docs.

File: pyproject.toml [hunk 1]
```diff
@@ -419,6 +402,17 @@ known-first-party = ["astropy", "extension_helpers"]
...
 
+[tool.ruff.format]
+exclude = ["examples/*",
+          "docs/*",
+          "docs/wcs/examples/*",
+          "docs/nddata/examples/*",
+          "cextern/*",
+          "astropy/extern/*",
+          "astropy/coordinates/.*tab.py",
+          "astropy/units/format/.*tab.py"]
```
Comment (eerovaher) [inline pyproject.toml:413]: Ruff formatter settings should precede Ruff linter settings to maintain alphabetical ordering. Furthermore, the configuration can be simplified to
```suggestion
[tool.ruff.format]
exclude = ["docs/*", "examples/*"]
```

**Review by eerovaher — CHANGES_REQUESTED** [review]

File: astropy/time/formats.py [hunk 2]
```diff
@@ -698,7 +698,10 @@ def set_jds(self, val1, val2):
     def to_value(self, **kwargs):
         scale = self.scale.upper().encode("ascii")
         iy_start, ims, ids, ihmsfs = erfa.d2dtf(
-            scale, 0, self.jd1, self.jd2  # precision=0
+            scale,
+            0,
+            self.jd1,
+            self.jd2,  # precision=0
```
[the remark this answers is not in the document]
Comment (mhvk) [reply astropy/time/formats.py:704]: Maybe just move the comment to the line before so this line can collapse to a single one again (be sure to remove trailing `,` on last item - or does the `ruff` version not care about that?

File: astropy/modeling/tests/test_models_quantities.py [hunk 3]
```diff
@@ -809,12 +809,12 @@ def test_models_fitting(model, fitter):
...
     if len(model["evaluation"][0]) == 2:
         x = np.linspace(1, 3, 100) * model["evaluation"][0][0].unit
-        y = np.exp(-x.value**2) * model["evaluation"][0][1].unit
+        y = np.exp(-(x.value**2)) * model["evaluation"][0][1].unit
         args = [x, y]
     else:
         x = np.linspace(1, 3, 100) * model["evaluation"][0][0].unit
         y = np.linspace(1, 3, 100) * model["evaluation"][0][1].unit
-        z = np.exp(-x.value**2 - y.value**2) * model["evaluation"][0][2].unit
+        z = np.exp(-(x.value**2) - y.value**2) * model["evaluation"][0][2].unit
```
Comment (mhvk) [inline astropy/modeling/tests/test_models_quantities.py:817]: why are those parentheses added (and not around `y.value` - this just makes the expression considerably *less* clear! Does adding a space after the `-` resolve it? If not, at least write `np.exp(-(x.value**2 + y.value**2))` for now. And please raise an issue upstream - this is just inane - math precedence rules exist for a reason.

File: astropy/modeling/tests/test_models_quantities.py [hunk 4]
```diff
@@ -809,12 +809,12 @@ def test_models_fitting(model, fitter):
     m = model["class"](**model["parameters"])
     if len(model["evaluation"][0]) == 2:
         x = np.linspace(1, 3, 100) * model["evaluation"][0][0].unit
-        y = np.exp(-x.value**2) * model["evaluation"][0][1].unit
+        y = np.exp(-(x.value**2)) * model["evaluation"][0][1].unit
```
Comment (mhvk) [inline astropy/modeling/tests/test_models_quantities.py:812]: As for the line below - though this is marginally bearable.

File: astropy/utils/tests/test_data.py [hunk 5]
```diff
@@ -1146,9 +1146,12 @@ def test_read_unicode(filename):
     contents = get_pkg_data_contents(os.path.join("data", filename), encoding="binary")
     assert isinstance(contents, bytes)
     x = contents.splitlines()[1]
-    assert x == (
-        b"\xff\xd7\x94\xd7\x90\xd7\xa1\xd7\x98\xd7\xa8\xd7\x95\xd7\xa0\xd7\x95"
-        b"\xd7\x9e\xd7\x99 \xd7\xa4\xd7\x99\xd7\x99\xd7\xaa\xd7\x95\xd7\x9f"[1:]
+    assert (
```
Comment (mhvk) [inline astropy/utils/tests/test_data.py:1149]: Hmm, one example where the change makes it less nice. Can one get rid of the inner parentheses?

File: pyproject.toml (hunk 1 above)
Comment (mhvk) [reply pyproject.toml:413]: `extern` is already excluded elsewhere, right?

**Review by mhvk — COMMENTED** [review]
Mostly changes for the better, except for two mathematics one. I think that should be reported upstream as undesirable.

File: astropy/modeling/tests/test_models_quantities.py (hunk 3 above)
Comment (eerovaher) [reply astropy/modeling/tests/test_models_quantities.py:817]: The `-` in front of `x.value` is different from the `-` in front of `y.value`. The former is a unary minus (`__neg__()`), whereas the latter is a subtraction (`__sub__()`). In Python exponentiation has higher precedence than unary minus, so `-x.value**2` is equivalent to `-(x.value**2)`, not `(-x.value)**2`, but apparently Ruff formatter thinks it's better to be explicit.

I agree that `np.exp(-(x.value**2 + y.value**2))` would be better.

File: pyproject.toml (hunk 1 above)
Comment (eerovaher) [reply pyproject.toml:413]: https://github.com/astropy/astropy/blob/aeeed0886ebfbc7642042e55920d62a82f3feff7/pyproject.toml#L323-L329

File: astropy/modeling/tests/test_models_quantities.py (hunk 3 above)
Comment (mhvk) [reply astropy/modeling/tests/test_models_quantities.py:817]: Yes, that makes sense -- but doesn't make the ruff formatter right!

File: pyproject.toml [hunk 6]
```diff
@@ -376,8 +359,15 @@ lint.ignore = [  # NOTE: non-permanent exclusions should be added to `.ruff.toml
 
     # Ruff-specific rules (RUF)
     "RUF005",  # unpack-instead-of-concatenating-to-collection-literal -- it's not clearly faster.
+
+    # ISC001 shouldn't be used with ruff format
+    # https://docs.astral.sh/ruff/formatter/#conflicting-lint-rules
+    "ISC001",
```
Comment (eerovaher) [inline pyproject.toml:365]: It's best to keep the list of ignored rules in alphabetical order. Admittedly RUF005 is in the wrong place too, but this pull request shouldn't worry about that (and I'm not convinced that RUF005 belongs on this list anyways).

**Review by eerovaher — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 6 above)
Comment (MridulS) [reply pyproject.toml]: Done! Well I would argue as soon as we want to keep the things in order here, we should have a formatter tool :)
Time for pyproject-fmt?

Comment (neutrinoceros) [reply pyproject.toml]: #15365  👀

**Review by eerovaher — APPROVED** [review]
This is now ready to be squashed and merged.

**Comment by pllim** [comment]
Not so fast, 3 warnings in doc build:

```
docs/development/codeguide.rst:575: WARNING: Duplicate explicit target name: "ruff".
docs/development/codeguide.rst:168: ERROR: Duplicate target name, cannot be used as a unique reference: "ruff".
docs/development/codeguide.rst:172: ERROR: Duplicate target name, cannot be used as a unique reference: "ruff".
```

**Review by pllim — APPROVED** [review]
Thanks!

**Comment by pllim** [comment]
7 approvals! This must be a record. 😸

**Pull request merged**